hypershift-operator/controllers/awsendpointservice/awsendpointservice_controller.go

```diff
@@ -0,0 +1,208 @@
+package awsendpointservice
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/aws/aws-sdk-go/service/elbv2"
+	"github.com/aws/aws-sdk-go/service/elbv2/elbv2iface"
+	"github.com/go-logr/logr"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	awsutil "github.com/openshift/hypershift/cmd/infra/aws/util"
+	"github.com/openshift/hypershift/support/upsert"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	finalizer                              = "hypershift.openshift.io/finalizer"
+	endpointServiceDeletionRequeueDuration = time.Duration(5 * time.Second)
+)
+
+type AWSEndpointServiceReconciler struct {
+	client.Client
+	upsert.CreateOrUpdateProvider
+	Region      string
+	ec2Client   ec2iface.EC2API
+	elbv2Client elbv2iface.ELBV2API
+}
+
+func (r *AWSEndpointServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Only add this controller to the manager if we have a non-empty credentials file
+	credsFile := os.Getenv("AWS_SHARED_CREDENTIALS_FILE")
+	if len(credsFile) == 0 {
+		return nil
+	}
+	credsBytes, err := ioutil.ReadFile(credsFile)
+	if err != nil || len(credsBytes) == 0 {
+		return nil
+	}
+
+	_, err = ctrl.NewControllerManagedBy(mgr).
+		For(&hyperv1.AWSEndpointService{}).
+		Build(r)
+	if err != nil {
+		return fmt.Errorf("failed setting up with a controller manager: %w", err)
+	}
+
+	awsSession := awsutil.NewSession("hypershift-operator")
+	awsConfig := awsutil.NewConfig(credsFile, r.Region)
+	r.ec2Client = ec2.New(awsSession, awsConfig)
+	r.elbv2Client = elbv2.New(awsSession, awsConfig)
+
+	return nil
+}
+
+func (r *AWSEndpointServiceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

Comment:
I have AWS throttling and quota on the mind right now... this seems reactive to `AWSEndpointService` changes, how frequent are those in practice? I ask because various AWS API calls in `reconcileAWSEndpointService()` seem like they would be invoked on every reconcile.

Reply:
All fields in the `AWSEndpointService` spec are immutable and this controller is the only one that updates status.

We could just immediately bail on the reconciliation if `status.endpointServiceName` is set and not verify that the Endpoint Service actually exists in AWS. What do you think?

Reply:
If it's easy and reliable to do a diff which can short-circuit reconciliation (and thus the API calls) which we know behaves correctly, I guess I don't see why not preemptively optimize it

Reply:
We can't be sure there is an Endpoint Service in AWS without making the call.  I think I'm going to leave it in there for now.